lib/iris/fileformats/grib/__init__.py

```diff
@@ -941,18 +941,33 @@ def save_grib2(cube, target, append=False, callback=None, **kwargs):
     See also :func:`iris.io.save`.
 
     """
-    # grib file (this bit is common to the pp and grib savers...)
-    if isinstance(target, basestring):
-        grib_file = open(target, "ab" if append else "wb")
-        filename = target
-    elif hasattr(target, "write"):
-        if hasattr(target, "mode") and "b" not in target.mode:
-            raise ValueError("Target not binary")
-        filename = target.name if hasattr(target, 'name') else None
-        grib_file = target
+    def apply_callback(pairs):
+        if isinstance(target, basestring):
+            filename = target
+        else:
+            filename = getattr(target, 'name', None)
+        for cube, message in pairs:
+            message = iris.io.run_saver_callback(callback, cube, message,
+                                                 filename)
+            if message is not None:
+                yield message
+
+    if callback is None:
+        messages = as_messages(cube)
     else:
-        raise ValueError("Can only save grib to filename or writable")
+        pairs = as_pairs(cube)
+        messages = apply_callback(pairs)
```

Comment:
NB. This demonstrates how one could apply a callback using the underlying `as_pairs()` and `save()` functions, but since the whole point of this PR is to look at a deconstructed save process we might well revert the `callback` argument anyway.